// react-ions/lib/components/Badge/Badge.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _style = require('./style.scss');

var _style2 = _interopRequireDefault(_style);

var _bind = require('classnames/bind');

var _bind2 = _interopRequireDefault(_bind);

var _Icon = require('../Icon');

var _Icon2 = _interopRequireDefault(_Icon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Badge = function Badge(props) {
  var cx = _bind2.default.bind(_style2.default);
  var iconPlusText = props.icon && props.text ? 'padded' : null;
  var iconSize = props.size === 'heavy' ? '30' : '16';
  var badgeClasses = cx(_style2.default.badge, props.theme, props.size, props.optClass, iconPlusText);

  return _react2.default.createElement(
    'div',
    { className: badgeClasses },
    props.icon ? _react2.default.createElement(_Icon2.default, { name: props.icon, height: iconSize, width: iconSize }) : null,
    props.text ? _react2.default.createElement(
      'span',
      null,
      props.text
    ) : null
  );
};

Badge.propTypes = {
  /**
   * Background color of the badge.
   */
  theme: _propTypes2.default.string,
  /**
   * Icon value to display in the badge.
   */
  icon: _propTypes2.default.string,
  /**
   * Text value to display in the badge (number or string).
   */
  text: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]),
  /**
   * The size of the badge.
   */
  size: _propTypes2.default.string,
  /**
   * Optional CSS class to pass to the badge.
   */
  optClass: _propTypes2.default.string
};

exports.default = Badge;
